x-pack/filebeat/input/etw/input.go

```diff
@@ -0,0 +1,215 @@
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/elastic-agent-libs/mapstr"
+)
+
+const (
+	inputName = "etw"
+)
+
+// etwInput struct holds the configuration and state for the ETW input
+type etwInput struct {
+	log        *logp.Logger
+	config     config
+	etwSession etw.Session
+}
+
+func Plugin() input.Plugin {
+	return input.Plugin{
+		Name:      inputName,
+		Stability: feature.Beta,
+		Info:      "Collect ETW logs.",
+		Manager:   stateless.NewInputManager(configure),
+	}
+}
+
+func configure(cfg *conf.C) (stateless.Input, error) {
+	conf := defaultConfig()
+	if err := cfg.Unpack(&conf); err != nil {
+		return nil, err
+	}
+
+	return &etwInput{
+		config: conf,
+	}, nil
+}
+
+func (e *etwInput) Name() string { return inputName }
+
+func (e *etwInput) Test(_ input.TestContext) error {
+	// ToDo
+	return nil
+}
+
+// Run starts the ETW session and processes incoming events.
+func (e *etwInput) Run(ctx input.Context, publisher stateless.Publisher) error {
+	var err error
+	// Initialize a new ETW session with the provided configuration.
+	e.etwSession, err = etw.NewSession(convertConfig(e.config))
+	if err != nil {
+		return fmt.Errorf("error when initializing '%s' session: %w", e.etwSession.Name, err)
+	}
+
+	// Set up logger with session information.
+	e.log = ctx.Logger.With("session", e.etwSession.Name)
+	e.log.Info("Starting " + inputName + " input")
+
+	var wg sync.WaitGroup
+	var once sync.Once
+
+	// Handle realtime session creation or attachment.
+	if e.etwSession.Realtime {
+		if !e.etwSession.NewSession {
+			// Attach to an existing session.
+			err = e.etwSession.GetHandler()
+			if err != nil {
+				return fmt.Errorf("unable to retrieve handler: %w", err)
+			}
+			e.log.Debug("attached to existing session")
+		} else {
+			// Create a new realtime session.
+			err = e.etwSession.CreateRealtimeSession()
+			if err != nil {
+				return fmt.Errorf("realtime session could not be created: %w", err)
+			}
+			e.log.Debug("created session")
+		}
+	}
+	// Defer the cleanup and closing of resources.
+	defer func() {
+		wg.Wait() // Ensure all goroutines have finished before closing.
+		once.Do(e.Close)
+		e.log.Info(inputName + " input stopped")
+	}()
+
+	// eventReceivedCallback processes each ETW event.
+	eventReceivedCallback := func(er *etw.EventRecord) uintptr {
+		if er == nil {
+			e.log.Error("received null event record")
+			return 1
+		}
+
+		e.log.Debugf("received event %d with length %d", er.EventHeader.EventDescriptor.Id, er.UserDataLength)
+
+		var event map[string]interface{}
+
+		if data, err := etw.GetEventProperties(er); err == nil {
+			event = map[string]interface{}{
+				"Header":          er.EventHeader,
+				"EventProperties": data,
+				"Metadata":        fillEventMetadata(er, e.etwSession, e.config),
+			}
+		} else {
+			e.log.Errorf("failed to read event properties: %w", err)
+			return 1
+		}
+
+		evt := beat.Event{
+			Timestamp: time.Now(),
+			Fields: mapstr.M{
```

Comment:
Is there an example event somewhere? I'm concerned that we are missing mappings for these fields and that they don't conform to existing conventions like being lower snake case.

Reply:
I updated the sample event in the PR description. Regarding the mapping for the event fields, I assumed that is a responsibility for further integrations to adapt incoming fields to ECS and normalise them (e.g. lowercase them). Maybe I misunderstood your comment, because there is no way to know in advance which fields are coming in which event.

Reply:
The input is being advertised in the docs for Filebeat, and it will be usable with standalone Filebeat which ships mappings for the `filebeat-*` indices. The fields from this input should be included. As an example this file has the aws-s3 mappings:

https://github.com/elastic/beats/blob/84e668cb1f75147d92c34b5d8e1e6207f2777e66/x-pack/filebeat/input/awss3/_meta/fields.yml

> because there is no way to know in advance which fields are coming in which event.

This is true of many inputs. The most common approach is to apply a dynamic mapping rule. For example with the `winlog` input all fields under `winlog.event_data.*` will receive a `keyword` field type mapping. But this input also has static fields that are known too like the process ID.

https://github.com/elastic/beats/blob/8bb171b4e17bd1dc7346adb21b6f412732051794/winlogbeat/_meta/fields.common.yml#L66-L75

Reply:
There are two approaches we could take here IMO. Either use a dedicated namespace like `etw` or make it look more like Windows Event Log data. For the latter I think everything that's in `winlog` now belongs in `winlog.event_data.*` and what you have in the header could mostly be placed in `winlog.*` where the fields have the same meanings.